Publish draft of npm best practices guide

jeffmendoza opened the pull request:
Step 2 of process.md. Publishing this draft to the drafts/ directory for
further refinement before proposal. Closes #2

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+   assess its trustworthiness through, for example, its number of contributors,
+   stars, etc.
+
+    1. Note: Non-ASCII characters and uppercase ASCII characters are [not
```
uppercase characters are supported for historical packages, but not for new ones, so it would be incorrect for tooling to ban them entirely.

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+- Mitigating certain threats such as malicious dependencies. A recent example
+  is the [color package's
+  sabotage](https://blog.sonatype.com/npm-libraries-colors-and-faker-sabotaged-in-protest-by-their-maintainer-what-to-do-now):
+  without a reproducible installation, developers installed and ran a newly
```
this seems disingenuous - it wouldn't have mattered if it was reproducible, because the developer intentionally sabotaged the project.

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+  published (compromised) version of the dependency on their CI/CD system,
+  giving an attacker immediate code execution on their infrastructure.
+
+- Detecting package corruptions before installation, for example due to RAM
```
i believe npm already does this by default, using an integrity hash

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+`integrity` field is not present), the installation will patch the lock file to
+reflect the installation.
+
+The lock file ***cannot*** be uploaded to a registry, which means that
+consumers who locally install the package via `npm install` may [see different
+dependency
+versions](https://dev.to/saurabhdaware/but-what-the-hell-is-package-lock-json-b04)
+than the repo owners used during testing. Using package-lock.json is akin to
+dynamic linking for low-level programming languages: the loader will resolve
+dependencies at load time using libraries available on the system. Using this
+lock file leaves the task of deciding dependencies' versions to use to the
+package consumer.
```
this fact is why i believe it's a *harmful* practice for published packages to use dev-only lockfiles: because it gives them a falsely insulated experience, while their users are exposed to breakage and vulns in transitive deps.

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+
+##### Shrinkwrap.json
+
+[Shrinkwrap.json](https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#package-lockjson-vs-npm-shrinkwrapjson)
+is another lock file supported by npm. The main difference is that this lock
+file ***may*** be uploaded to a registry along with the package. This ensures that
+consumers of the dependency will obtain the same dependencies' hashes as the
+repo owners intended. With shrinkwrap.json, the package producer takes
+responsibility for updating the dependencies on behalf of their consumers. It's
+akin to static linking for low-level programming languages: everything is
+declared at packaging time.
+
+To generate the shrinkwrap.json, an existing package-lock.json must be present
+and the command [`npm
+shrinkwrap`](https://docs.npmjs.com/cli/v8/commands/npm-shrinkwrap) must be
+run.
```
it's considered incredibly user-hostile to include a shrinkwrap file with a published package, because it blocks deduplication and *also* because it blocks the free flow of bugfixes and nonbreaking updates. This should be noted lest someone think this is a viable path for a package.

laurentsimon replied:
we're not recommending it, just documenting that it exists. As @jeffmendoza said, this document will go go thru an embargo period to let folks comment. So we'll be able to tweak the language and point out important limitations.

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+
+**Recommendations:**
+
+1. Developers should declare and commit a manifest file for ***all*** their
```
i strongly disagree with this recommendation, for the reasons given above, and none of my 300+ packages (which represent well over 10% of npm's download traffic) include a lockfile, nor will they, because I think that would cause far more harm than it helps.

laurentsimon replied:
here we don't mean shrinkwrap files, we just mean a lock file used for _CI_ - which is not used during package installation and won't "blocks the free flow of bugfixes and nonbreaking updates". 

I agree the way it's written is confusing. Thanks for pointing this out.

ljharb replied:
Yes, but I still think that such a dev-only lockfile is harmful, because it falsely insulates maintainers from problems their users will encounter.

ljharb commented on drafts/npm.md:
```diff
@@ -0,0 +1,472 @@
+
+**Recommendations:**
+
+1. Use an [Automation
+   token](https://docs.npmjs.com/creating-and-viewing-access-tokens) to
+   authenticate and publish to the default npm registry.
```
this is strictly less secure than a human doing it locally, because it's a single factor, instead of the two-factor that human publishes require.

laurentsimon replied:
Good point. Let's split this into different sections for "manual" and "automated".

jeffmendoza commented:
@ljharb fyi We'll be presenting the [process](https://github.com/ossf/package-manager-best-practices/blob/main/process.md) in tomorrow morning's WG meeting for how a doc will go from draft to rfc to published. At this point, I'm converting from a Google doc to markdown, and am not addressing comments. Once we have it in `drafts/` comments can be opened as issues, but the main period where authors will be addressing comments is during the 30-day review period. Thanks.

ljharb commented:
ok, thanks for clarifying.

jeffmendoza commented:
@laurentsimon Should all be here, ptal. I'll move over the existing comments as new issues once merged.

laurentsimon reviewed (approved):
Thanks!
